impeller/entity/save_layer_utils.cc

```diff
@@ -0,0 +1,75 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/entity/save_layer_utils.h"
+
+namespace impeller {
+
+std::optional<Rect> ComputeSaveLayerCoverage(
+    const Rect& content_coverage,
+    const Matrix& effect_transform,
+    const Rect& coverage_limit,
+    const std::shared_ptr<FilterContents>& image_filter,
+    bool flood_output_coverage,
+    bool flood_input_coverage) {
+  Rect coverage = content_coverage;
+  // There are two conditions that currently trigger flood_input_coverage, the
+  // first is the presence of a backdrop filter and the second is the presence
+  // of a color filter that effects transparent black.
+  //
+  // Backdrop filters apply before the saveLayer is restored. The presence of
+  // a backdrop filter causes the content coverage of the saveLayer to be
+  // unbounded.
+  //
+  // If there is a color filter that needs to flood its output. The color filter
+  // is applied before any image filters, so this floods input coverage and not
+  // the output coverage. Technically, we only need to flood the output of the
+  // color filter and could allocate a render target sized just to the content,
+  // but we don't currenty have the means to do so. Flooding the coverage is a
+  // non-optimal but technically correct way to render this.
+  if (flood_input_coverage) {
+    coverage = Rect::MakeMaximum();
+  }
+
+  // If flood_output_coverage is true, then the restore is applied with a
+  // destructive blend mode that requires flooding to the coverage limit.
+  // Technically we could only allocated a render target as big as the input
+  // coverage and then use a decal sampling mode to perform the flood. Returning
+  // the coverage limit is a correct but non optimal means of ensuring correct
+  // rendering.
```

Comment:
But, this return value will be used to control the allocation of an RT for the content of this SL. If there is a transforming IF then the coverage_limit might not be enough for that case.

Consider an IF that scales down by 50%. If you want to collect pixels so that they will be filtered to produce coverage_limit, then you need to collect 2x pixels in both directions, so you need coverage_limit*2 for the RT, which will then be processed by the IF to create pixels that will fill the coverage_limit.

Reply:
Ahh right, I was assuming that the IF could only make this smaller. But it could make it bigger too. If this case is true what I actually want is to return the source_coverage_limit instead.

Reply:
I mean only add part of it back

Reply:
woops wrong message